core/ledger/pvtdatastorage/retroactive_hashed_index.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package pvtdatastorage
+
+import (
+	"github.com/hyperledger/fabric-protos-go/ledger/rwset"
+	"github.com/hyperledger/fabric/common/ledger/util/leveldbhelper"
+
+	"github.com/hyperledger/fabric/core/ledger"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/rwsetutil"
+	"github.com/hyperledger/fabric/core/ledger/util"
+	"github.com/pkg/errors"
+)
+
+const (
+	previousDataVersion = ""
+	currentDataVersion  = "2.6"
+	maxUpgradeBatchSize = 4 * 1024 * 1024 // 4 MB
+)
+
+func checkAndConstructHashedIndex(storePath string, ledgerIDs []string) error {
+	i, err := leveldbhelper.RetrieveDataFormatInfo(storePath)
+	if err != nil {
+		return err
+	}
+
+	if i.IsDBEmpty || i.FormatVerison == currentDataVersion {
+		return nil
+	}
+
+	if i.FormatVerison == previousDataVersion {
+		if err := constructHashedIndex(storePath, ledgerIDs); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	return errors.Errorf("cannot upgrade data format for pvtdatastore from %s to %s", i.FormatVerison, currentDataVersion)
```

Comment:
Shouldn't the error be something like "unexpected version" rather than "cannot upgrade".

Reply:
Agreed!